File: app/src/main/java/com/zwstudio/logicpuzzlesandroid/puzzles/minilits/domain/MiniLitsGameState.java
package com.zwstudio.logicpuzzlesandroid.puzzles.minilits.domain;

import com.zwstudio.logicpuzzlesandroid.common.domain.AllowedObjectState;
import com.zwstudio.logicpuzzlesandroid.common.domain.CellsGameState;
import com.zwstudio.logicpuzzlesandroid.common.domain.Graph;
import com.zwstudio.logicpuzzlesandroid.common.domain.MarkerOptions;
import com.zwstudio.logicpuzzlesandroid.common.domain.Node;
import com.zwstudio.logicpuzzlesandroid.common.domain.Position;
import com.zwstudio.logicpuzzlesandroid.home.domain.HintState;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

import fj.F;
import fj.Ord;
import fj.data.Option;
import fj.function.Effect1;

import static fj.data.Array.array;
import static fj.data.HashMap.fromMap;
import static fj.data.List.iterableList;

/**
 * Created by zwvista on 2016/09/29.
 */

public class MiniLitsGameState extends CellsGameState<MiniLitsGame, MiniLitsGameMove, MiniLitsGameState> {
    public MiniLitsObject[] objArray;
    public Map<Position, HintState> pos2state = new HashMap<>();

    public MiniLitsGameState(MiniLitsGame game) {
        super(game);
        objArray = new MiniLitsObject[rows() * cols()];
        for (int i = 0; i < objArray.length; i++)
            objArray[i] = new MiniLitsEmptyObject();
        updateIsSolved();
    }

    public MiniLitsObject get(int row, int col) {
        return objArray[row * cols() + col];
    }
    public MiniLitsObject get(Position p) {
        return get(p.row, p.col);
    }
    public void set(int row, int col, MiniLitsObject dotObj) {
        objArray[row * cols() + col] = dotObj;
    }
    public void set(Position p, MiniLitsObject obj) {
        set(p.row, p.col, obj);
    }

    private class MiniLitsAreaInfo {
        List<Position> trees = new ArrayList<>();
        Set<Integer> blockIndexes = new HashSet<>();
        Set<Integer> neighborIndexes = new HashSet<>();
        int tetrominoIndex = -1;
    }

    public boolean setObject(MiniLitsGameMove move) {
        if (!isValid(move.p) || get(move.p).equals(move.obj)) return false;
        set(move.p, move.obj);
        updateIsSolved();
        return true;
    }

    public boolean switchObject(MiniLitsGameMove move) {
        MarkerOptions markerOption = MarkerOptions.values()[game.gdi.getMarkerOption()];
        F<MiniLitsObject, MiniLitsObject> f = obj -> {
            if (obj instanceof MiniLitsEmptyObject)
                return markerOption == MarkerOptions.MarkerFirst ?
                        new MiniLitsMarkerObject() : new MiniLitsTreeObject();
            if (obj instanceof MiniLitsTreeObject)
                return markerOption == MarkerOptions.MarkerLast ?
                        new MiniLitsMarkerObject() : new MiniLitsEmptyObject();
            if (obj instanceof MiniLitsMarkerObject)
                return markerOption == MarkerOptions.MarkerFirst ?
                        new MiniLitsTreeObject() : new MiniLitsEmptyObject();
            return obj;
        };
        MiniLitsObject o = get(move.p);
        move.obj = f.f(o);
        return setObject(move);
    }

    /*
        iOS Game: Logic Games/Puzzle Set 14/Mini-Lits

        Summary
        Lits Jr.

        Description
        1. You play the game with triominos (pieces of three squares).
        2. The board is divided into many areas. You have to place a triomino
           into each area respecting these rules:
        3. No two adjacent (touching horizontally / vertically) triominos should
           be of equal shape & orientation.
        4. All the shaded cells should form a valid Nurikabe.
    */
    private void updateIsSolved() {
        boolean allowedObjectsOnly = game.gdi.isAllowedObjectsOnly();
        isSolved = true;
        Graph g = new Graph();
        Map<Position, Node> pos2node = new HashMap<>();
        for (int r = 0; r < rows(); r++)
            for (int c = 0; c < cols(); c++) {
                Position p = new Position(r, c);
                MiniLitsObject o = get(p);
                if (o instanceof MiniLitsForbiddenObject)
                    set(r, c, new MiniLitsEmptyObject());
                else if(o instanceof MiniLitsTreeObject) {
                    MiniLitsTreeObject o2 = (MiniLitsTreeObject)o;
                    o2.state = AllowedObjectState.Normal;
                    Node node = new Node(p.toString());
                    g.addNode(node);
                    pos2node.put(p, node);
                }
            }
        for (Map.Entry<Position, Node> entry : pos2node.entrySet()) {
            Position p = entry.getKey();
            Node node = entry.getValue();
            for (Position os : MiniLitsGame.offset) {
                Position p2 = p.add(os);
                Node node2 = pos2node.get(p2);
                if (node2 != null) g.connectNode(node, node2);
            }
        }
        List<List<Position>> blocks = new ArrayList<>();
        while (!pos2node.isEmpty()) {
            g.setRootNode(iterableList(pos2node.values()).head());
            List<Node> nodeList = g.bfs();
            List<Position> block = fromMap(pos2node).toStream().filter(e -> nodeList.contains(e._2())).map(e -> e._1()).toJavaList();
            blocks.add(block);
            for (Position p : block)
                pos2node.remove(p);
        }
        // 4. All the shaded cells should form a valid Nurikabe.
        if (blocks.size() != 1) isSolved = false;
        List<MiniLitsAreaInfo> infos = new ArrayList<>();
        for (int i = 0; i < game.areas.size(); i++)
            infos.add(new MiniLitsAreaInfo());
        for (int i = 0; i < blocks.size(); i++) {
            List<Position> block = blocks.get(i);
            for (Position p : block) {
                Integer n = game.pos2area.get(p);
                MiniLitsAreaInfo info = infos.get(n);
                info.trees.add(p);
                info.blockIndexes.add(i);
            }
        }
        for (int i = 0; i < infos.size(); i++) {
            MiniLitsAreaInfo info = infos.get(i);
            for (Position p : info.trees)
                for (Position os : MiniLitsGame.offset) {
                    Position p2 = p.add(os);
                    Option<Integer> index = iterableList(infos).toStream().indexOf(info2 -> info2.trees.contains(p2));
                    if (index.isSome() && index.some() != i)
                        info.neighborIndexes.add(index.some());
                }
        }
        Effect1<MiniLitsAreaInfo> notSolved = info -> {
            isSolved = false;
            for (Position p : info.trees) {
                MiniLitsTreeObject o = (MiniLitsTreeObject)get(p);
                o.state = AllowedObjectState.Error;
            }
        };
        for (int i = 0; i < infos.size(); i++) {
            MiniLitsAreaInfo info = infos.get(i);
            int treeCount = info.trees.size();
            if (treeCount >= 3 && allowedObjectsOnly)
                for (Position p : game.areas.get(i)) {
                    MiniLitsObject o = get(p);
                    if (o instanceof MiniLitsEmptyObject || o instanceof MiniLitsMarkerObject)
                        set(p, new MiniLitsForbiddenObject());
                }
            if (treeCount > 3 || treeCount == 3 && info.blockIndexes.size() > 1) notSolved.f(info);
            // 2. The board is divided into many areas. You have to place a triomino
            // into each area.
            if (treeCount == 3 && info.blockIndexes.size() == 1) {
                Collections.sort(info.trees, Position::compareTo);
                List<Position> treeOffsets = new ArrayList<>();
                Position p2 = new Position(iterableList(info.trees).map(p -> p.row).minimum(Ord.intOrd),
                        iterableList(info.trees).map(p -> p.col).minimum(Ord.intOrd));
                for (Position p : info.trees)
                    treeOffsets.add(p.subtract(p2));
                info.tetrominoIndex = array(MiniLitsGame.triominos).toStream()
                        .indexOf(arr -> Arrays.equals(arr, treeOffsets.toArray())).orSome(-1);
                if (info.tetrominoIndex == -1) notSolved.f(info);
            }
            if (treeCount < 3) isSolved = false;
        }
        // 3. No two adjacent (touching horizontally / vertically) triominos should
        // be of equal shape & orientation.
        for (int i = 0; i < infos.size(); i++) {
            MiniLitsAreaInfo info = infos.get(i);
            int index = info.tetrominoIndex;
            if (index == -1) continue;
            if (iterableList(info.neighborIndexes).exists(j -> infos.get(j).tetrominoIndex == index)) notSolved.f(info);
        }
        if (!isSolved) return;
        // 4. All the shaded cells should form a valid Nurikabe.
        List<Position> block = blocks.get(0);
        rule2x2: for (Position p : block) {
            for (Position os : MiniLitsGame.offset3)
                if (block.contains(p.add(os)))
                    continue rule2x2;
            isSolved = false;
            for (Position os : MiniLitsGame.offset3) {
                MiniLitsTreeObject o = new MiniLitsTreeObject();
                o.state = AllowedObjectState.Error;
                set(p.add(os), o);
            }
        }
    }
}

File: app/src/main/java/com/zwstudio/logicpuzzlesandroid/puzzles/minilits/domain/MiniLitsGameState.kt
package com.zwstudio.logicpuzzlesandroid.puzzles.minilits.domainimport

import com.zwstudio.logicpuzzlesandroid.common.domain.*
import com.zwstudio.logicpuzzlesandroid.home.domain.HintState
import fj.F
import fj.data.Stream
import java.util.*

class MiniLitsGameState(game: MiniLitsGame?) : CellsGameState<MiniLitsGame?, MiniLitsGameMove?, MiniLitsGameState?>(game) {
    var objArray: Array<MiniLitsObject?>
    var pos2state: Map<Position, HintState> = HashMap<Position, HintState>()
    operator fun get(row: Int, col: Int) = objArray[row * cols() + col]

    operator fun get(p: Position) = get(p!!.row, p.col)

    operator fun set(row: Int, col: Int, dotObj: MiniLitsObject?) {
        objArray[row * cols() + col] = dotObj
    }

    operator fun set(p: Position, obj: MiniLitsObject?) {
        set(p!!.row, p.col, obj)
    }

    private inner class MiniLitsAreaInfo {
        var trees = mutableListOf<Position>()
        var blockIndexes = mutableSetOf<Int>()
        var neighborIndexes = mutableSetOf<Int>()
        var tetrominoIndex = -1
    }

    fun setObject(move: MiniLitsGameMove): Boolean {
        if (!isValid(move.p) || get(move.p) == move.obj) return false
        set(move.p, move.obj)
        updateIsSolved()
        return true
    }

    fun switchObject(move: MiniLitsGameMove): Boolean {
        val markerOption: MarkerOptions = MarkerOptions.values().get(game.gdi.getMarkerOption())
        val f: F<MiniLitsObject, MiniLitsObject> = label@ F<MiniLitsObject, MiniLitsObject> { obj: MiniLitsObject? ->
            if (obj is MiniLitsEmptyObject) return@label if (markerOption == MarkerOptions.MarkerFirst) MiniLitsMarkerObject() else MiniLitsTreeObject()
            if (obj is MiniLitsTreeObject) return@label if (markerOption == MarkerOptions.MarkerLast) MiniLitsMarkerObject() else MiniLitsEmptyObject()
            if (obj is MiniLitsMarkerObject) return@label if (markerOption == MarkerOptions.MarkerFirst) MiniLitsTreeObject() else MiniLitsEmptyObject()
            obj
        }
        val o: MiniLitsObject? = get(move.p)
        move.obj = f.f(o)
        return setObject(move)
    }

    /*
        iOS Game: Logic Games/Puzzle Set 14/Mini-Lits

        Summary
        Lits Jr.

        Description
        1. You play the game with triominos (pieces of three squares).
        2. The board is divided into many areas. You have to place a triomino
           into each area respecting these rules:
        3. No two adjacent (touching horizontally / vertically) triominos should
           be of equal shape & orientation.
        4. All the shaded cells should form a valid Nurikabe.
    */
    private fun updateIsSolved() {
        val allowedObjectsOnly: Boolean = game.gdi.isAllowedObjectsOnly()
        isSolved = true
        val g = Graph()
        val pos2node = mutableMapOf<Position, Node>()
        for (r in 0 until rows()) for (c in 0 until cols()) {
            val p = Position(r, c)
            val o: MiniLitsObject? = get(p)
            if (o is MiniLitsForbiddenObject) set(r, c, MiniLitsEmptyObject()) else if (o is MiniLitsTreeObject) {
                o.state = AllowedObjectState.Normal
                val node = Node(p.toString())
                g.addNode(node)
                pos2node[p] = node
            }
        }
        for ((p, node) in pos2node) {
            for (os in MiniLitsGame.Companion.offset) {
                val p2 = p.add(os)
                val node2 = pos2node[p2]
                if (node2 != null) g.connectNode(node, node2)
            }
        }
        val blocks = mutableListOf<List<Position>>()
        while (!pos2node.isEmpty()) {
            g.setRootNode(fj.data.List.iterableList(pos2node.values).head())
            val nodeList = g.bfs()
            val block = fj.data.HashMap.fromMap(pos2node).toStream().filter(F<P2<Position, Node>, Boolean> { e: P2<Position?, Node?> -> nodeList.contains(e._2()) }).map<Position>(F<P2<Position, Node>, Position> { e: P2<Position?, Node?> -> e._1() }).toJavaList()
            blocks.add(block)
            for (p in block) pos2node.remove(p)
        }
        // 4. All the shaded cells should form a valid Nurikabe.
        if (blocks.size != 1) isSolved = false
        val infos = Stream.range(0, game.areas.size.toLong()).map<MiniLitsAreaInfo>(F<Int, MiniLitsAreaInfo> { i: Int? -> MiniLitsAreaInfo() }).toJavaList()
        for (i in blocks.indices) {
            val block = blocks[i]
            for (p in block) {
                val n: Int = game.pos2area.get(p)
                val info = infos[n]
                info.trees.add(p)
                info.blockIndexes.add(i)
            }
        }
        for (i in infos.indices) {
            val info = infos[i]
            for (p in info.trees) for (os in MiniLitsGame.Companion.offset) {
                val p2 = p.add(os)
                val index = fj.data.List.iterableList(infos).toStream().indexOf(F<MiniLitsAreaInfo, Boolean> { info2: MiniLitsAreaInfo -> info2.trees.contains(p2) })
                if (index.isSome && index.some() != i) info.neighborIndexes.add(index.some())
            }
        }
        val notSolved: Effect1<MiniLitsAreaInfo> = Effect1<MiniLitsAreaInfo> { info: MiniLitsAreaInfo ->
            isSolved = false
            for (p in info.trees) {
                val o: MiniLitsTreeObject? = get(p) as MiniLitsTreeObject?
                o!!.state = AllowedObjectState.Error
            }
        }
        for (i in infos.indices) {
            val info = infos[i]
            val treeCount = info.trees.size
            if (treeCount >= 3 && allowedObjectsOnly) for (p in game.areas.get(i)) {
                val o: MiniLitsObject? = get(p)
                if (o is MiniLitsEmptyObject || o is MiniLitsMarkerObject) set(p, MiniLitsForbiddenObject())
            }
            if (treeCount > 3 || treeCount == 3 && info.blockIndexes.size > 1) notSolved.f(info)
            // 2. The board is divided into many areas. You have to place a triomino
            // into each area.
            if (treeCount == 3 && info.blockIndexes.size == 1) {
                Collections.sort(info.trees) { obj: Position, other: Position? -> obj.compareTo(other) }
                val treeOffsets = mutableListOf<Position>()
                val p2 = Position(fj.data.List.iterableList(info.trees).map<Int>(F<Position, Int> { p: Position -> p.row }).minimum(Ord.intOrd),
                    fj.data.List.iterableList(info.trees).map<Int>(F<Position, Int> { p: Position -> p.col }).minimum(Ord.intOrd))
                for (p in info.trees) treeOffsets.add(p.subtract(p2))
                info.tetrominoIndex = fj.data.Array.array<Array<Position>>(*MiniLitsGame.Companion.triominos).toStream()
                    .indexOf(F<Array<Position>, Boolean> { arr: Array<Position?>? -> Arrays.equals(arr, treeOffsets.toTypedArray()) }).orSome(-1)
                if (info.tetrominoIndex == -1) notSolved.f(info)
            }
            if (treeCount < 3) isSolved = false
        }
        // 3. No two adjacent (touching horizontally / vertically) triominos should
        // be of equal shape & orientation.
        for (i in infos.indices) {
            val info = infos[i]
            val index = info.tetrominoIndex
            if (index == -1) continue
            if (fj.data.List.iterableList(info.neighborIndexes).exists(F<Int, Boolean> { j: Int? -> infos[j!!].tetrominoIndex == index })) notSolved.f(info)
        }
        if (!isSolved) return
        // 4. All the shaded cells should form a valid Nurikabe.
        val block = blocks[0]
        rule2x2@ for (p in block) {
            for (os in MiniLitsGame.Companion.offset3) if (block.contains(p.add(os))) continue@rule2x2
            isSolved = false
            for (os in MiniLitsGame.Companion.offset3) {
                val o = MiniLitsTreeObject()
                o.state = AllowedObjectState.Error
                set(p.add(os), o)
            }
        }
    }

    init {
        objArray = arrayOfNulls<MiniLitsObject>(rows() * cols())
        for (i in objArray.indices) objArray[i] = MiniLitsEmptyObject()
        updateIsSolved()
    }
}